scipy/optimize/_numdiff.py

```diff
@@ -551,32 +551,43 @@ def _dense_difference(fun, x0, f0, h, use_one_sided, method):
     h_vecs = np.diag(h)
 
     for i in range(h.size):
+        df = np.zeros_like(f0)
+
         if method == '2-point':
             x = x0 + h_vecs[i]
             dx = x[i] - x0[i]  # Recompute dx as exactly representable number.
-            df = fun(x) - f0
+            if np.count_nonzero(dx):
+                df = fun(x) - f0
         elif method == '3-point' and use_one_sided[i]:
             x1 = x0 + h_vecs[i]
             x2 = x0 + 2 * h_vecs[i]
             dx = x2[i] - x0[i]
-            f1 = fun(x1)
-            f2 = fun(x2)
-            df = -3.0 * f0 + 4 * f1 - f2
+            if np.count_nonzero(dx):
+                f1 = fun(x1)
+                f2 = fun(x2)
+                df = -3.0 * f0 + 4 * f1 - f2
         elif method == '3-point' and not use_one_sided[i]:
             x1 = x0 - h_vecs[i]
             x2 = x0 + h_vecs[i]
             dx = x2[i] - x1[i]
-            f1 = fun(x1)
-            f2 = fun(x2)
-            df = f2 - f1
+            if np.count_nonzero(dx):
+                f1 = fun(x1)
+                f2 = fun(x2)
+                df = f2 - f1
         elif method == 'cs':
             f1 = fun(x0 + h_vecs[i]*1.j)
             df = f1.imag
             dx = h_vecs[i, i]
         else:
             raise RuntimeError("Never be here.")
 
-        J_transposed[i] = df / dx
+        with np.errstate(invalid='ignore'):
+            # filter warning for 0 / 0
+            # However a 1.0 / 0.0 will still raise a
+            # RuntimeWarning: divide by zero encountered in true_divide
+            g = np.asarray(df / dx)
```

Comment:
What will be result in `g` when `df` and `dx` are zeros? I thought `nan`, but apparently is `0`. 

Should it be documented, that in case `lb[i] == ub[i]` the corresponding column in Jacobian will be zero?

Reply:
You're right that in floating point arithmetic if `df` and `dx` are zeros then `df/dx` is `nan`. If people set `lb[i] == ub[i]` from a bounded optimizer that calculates gradients then this calculation will emit a RuntimeWarning:

```
RuntimeWarning: invalid value encountered in true_divide
  np.array([0.0]) / np.array([0.0])
```

Reply:
I saw `np.array(0) / np.array(0) = np.array(0)` (with a warning) in numpy 1.13 and Python 2.7.

I think `nan` is an expected result now (as it should be in my opinion). Not sure if it worth investigating in detail.

Reply:
I agree with you in principle that `approx_derivative` should return nan for equal bounds. But if `approx_derivative` returns a `nan` for those entries, and emits a RuntimeWarning, then the calling code in various `minimize` methods will need to:

- filter the RuntimeWarning, because users wouldn't expect a warning from setting `lb[i] == ub[i]`, which has historically been allowed.
- change all the nan entries in the gradient to 0 because the SLSQP/LBFGSB/etc Fortran code can't deal with nan entries in a grad matrix.

Overall we need to come up with a fix for #12502. There are several other duplicate issues, so it's a pain point.

Reply:
@nmayorov, these changes to `approx_derivative` will return `np.nan` if a lower bound is equal to an upper bound. I've noted that in the docstring.
Note: those np.nans are then converted to zero in `ScalarFunction`.